server/src/main/java/org/opensearch/tasks/TaskResourceTrackingService.java

```diff
@@ -0,0 +1,255 @@
+        "task_resource_tracking.enabled",
+        true,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    public static final String TASK_ID = "TASK_ID";
+
+    private static final ThreadMXBean threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+
+    private final ConcurrentMapLong<Task> resourceAwareTasks = ConcurrentCollections.newConcurrentMapLongWithAggressiveConcurrency();
+    private final ThreadPool threadPool;
+    private volatile boolean taskResourceTrackingEnabled;
+
+    @Inject
+    public TaskResourceTrackingService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool) {
+        this.taskResourceTrackingEnabled = TASK_RESOURCE_TRACKING_ENABLED.get(settings);
+        this.threadPool = threadPool;
+        clusterSettings.addSettingsUpdateConsumer(TASK_RESOURCE_TRACKING_ENABLED, this::setTaskResourceTrackingEnabled);
+    }
+
+    public void setTaskResourceTrackingEnabled(boolean taskResourceTrackingEnabled) {
+        this.taskResourceTrackingEnabled = taskResourceTrackingEnabled;
+    }
+
+    public boolean isTaskResourceTrackingEnabled() {
+        return taskResourceTrackingEnabled;
+    }
+
+    public boolean isTaskResourceTrackingSupported() {
+        return threadMXBean.isThreadAllocatedMemorySupported() && threadMXBean.isThreadAllocatedMemoryEnabled();
+    }
+
+    /**
+     * Executes logic only if task supports resource tracking and resource tracking setting is enabled.
+     * <p>
+     * 1. Starts tracking the task in map of resourceAwareTasks.
+     * 2. Adds Task Id in thread context to make sure it's available while task is processed across multiple threads.
+     *
+     * @param task for which resources needs to be tracked
+     * @return Autocloseable stored context to restore ThreadContext to the state before this method changed it.
+     */
+    public ThreadContext.StoredContext startTracking(Task task) {
+        if (task.supportsResourceTracking() == false
+            || isTaskResourceTrackingEnabled() == false
+            || isTaskResourceTrackingSupported() == false) {
+            return () -> {};
+        }
+
+        logger.debug("Starting resource tracking for task: {}", task.getId());
+        resourceAwareTasks.put(task.getId(), task);
+        return addTaskIdToThreadContext(task);
+    }
+
+    /**
+     * Stops tracking task registered earlier for tracking.
+     * <p>
+     * It doesn't have feature enabled check to avoid any issues if setting was disable while the task was in progress.
+     * <p>
+     * It's also responsible to stop tracking the current thread's resources against this task if not already done.
+     * This happens when the thread executing the request logic itself calls the unregister method. So in this case unregister
+     * happens before runnable finishes.
+     *
+     * @param task task which has finished and doesn't need resource tracking.
+     */
+    public void stopTracking(Task task) {
```

Comment:
Since this is always called from ``unregister `` of task manager, disabling of the resource tracking setting will only be reflective for the new tasks?

Reply:
Yes. For the tasks which were running before setting was disabled will continue to get resource info and only the new task will skip the logic.